setup.py

```diff
@@ -17,11 +19,194 @@
+        return dict(
+            map(lambda l: list(map(str.strip, l.split('=', 1))),
+                filter(bool, map(str.strip, str(v).strip().splitlines())))
+        )
+
+    def cfg_val_to_primitive(v):
+        """Parse primitive config val to appropriate data type."""
+        return json.loads(v.strip().lower())
+
+    def read_configuration(filepath):
+        """Read metadata and options from setup.cfg located at filepath."""
+        cfg = ConfigParser()
+        with io.open(filepath, encoding='utf-8') as f:
+            cfg.read_file(f)
+
+        md = dict(cfg.items('metadata'))
+        for list_key in 'classifiers', 'keywords':
+            try:
+                md[list_key] = cfg_val_to_list(md[list_key])
+            except KeyError:
+                pass
+        try:
+            md['long_description'] = maybe_read_files(md['long_description'])
+        except KeyError:
+            pass
+        opt = dict(cfg.items('options'))
+        for list_key in 'use_scm_version', 'zip_safe':
+            try:
+                opt[list_key] = cfg_val_to_primitive(opt[list_key])
+            except KeyError:
+                pass
+        for list_key in 'scripts', 'install_requires', 'setup_requires':
+            try:
+                opt[list_key] = cfg_val_to_list(opt[list_key])
+            except KeyError:
+                pass
+        try:
+            opt['package_dir'] = cfg_val_to_dict(opt['package_dir'])
+        except KeyError:
+            pass
+        try:
+            opt_package_data = dict(cfg.items('options.package_data'))
+            if not opt_package_data.get('', '').strip():
+                opt_package_data[''] = opt_package_data['*']
+                del opt_package_data['*']
+        except (KeyError, NoSectionError):
+            opt_package_data = {}
+        try:
+            opt_extras_require = dict(cfg.items('options.extras_require'))
+            opt['extras_require'] = {}
+            for k, v in opt_extras_require.items():
+                opt['extras_require'][k] = cfg_val_to_list(v)
+        except NoSectionError:
+            pass
+        opt['package_data'] = {}
+        for k, v in opt_package_data.items():
+            opt['package_data'][k] = cfg_val_to_list(v)
+        cur_pkgs = opt.get('packages', '').strip()
+        if '\n' in cur_pkgs:
+            opt['packages'] = cfg_val_to_list(opt['packages'])
+        elif cur_pkgs.startswith('find:'):
+            opt_packages_find = stringify_dict_contents(
+                dict(cfg.items('options.packages.find'))
+            )
+            opt['packages'] = setuptools.find_packages(**opt_packages_find)
+        return {'metadata': md, 'options': opt}
+
+
+def cut_local_version_on_upload(version):
+    """Generate a PEP440 local version if uploading to PyPI."""
+    import os
+    import setuptools_scm.version  # only present during setup time
+    IS_PYPI_UPLOAD = os.getenv('PYPI_UPLOAD') == 'true'
```

Comment:
Perhaps comment that this is set from the `tox.ini` to help later reviews?

Reply:
I'm thinking that I'll maybe extend the shim to support `attr: ` in `name` and will re-wire version generator as I did in another recent project of mine: https://github.com/sanitizers/octomachinery/blob/ce217ba/setup.cfg#L9.
This seems to be more explicit. Let's see if I can do it...

Reply:
Turned out to be more complicated than I thought... Maybe another time.